pkg/controller/infra/infra_controller.go

```diff
@@ -233,28 +525,176 @@ func (c *InfraController) syncCluster(key string) error {
 		return nil
 	}
 	if err != nil {
 		return err
 	}
 
-	// Deep-copy otherwise we are mutating our cache.
-	// TODO: Deep-copy only when needed.
-	cluster = cluster.DeepCopy()
+	if !c.expectations.SatisfiedExpectations(key) {
+		// expectations have not been met, come back later
+		cLog.Debugln("cluster has not satisfied expectations yet")
+		return nil
+	}
 
-	clusterLogger := c.logger.WithField("cluster", cluster.Name)
+	// Look through infra jobs that belong to this cluster.
+	// If the job's generation corresponds to the cluster's current generation
+	// sync the cluster status with the job's latest status.
+	// If the job is active but does not correspond to the cluster's current generation
+	// add it to the list of jobs to be terminated.
+	invalidActiveJobs := []*v1batch.Job{}
+	var currentInfraJob *v1batch.Job
+	jobs, err := c.jobsLister.Jobs(cluster.Namespace).List(labels.Everything())
+	if err != nil {
+		return err
+	}
+	for _, job := range jobs {
+		if isInfraJob(job) && isController(cluster, job) {
+			switch {
+			case jobClusterGeneration(job) == cluster.Generation:
+				currentInfraJob = job
+			case isActive(job):
+				cLog.WithField("job", jobKey(job)).
+					Debugln("found active job that does not correspond to cluster's current generation")
+				invalidActiveJobs = append(invalidActiveJobs, job)
+				break
+			}
+		}
+	}
+
+	// Determine expectations for creates and deletes
+	shouldProvision := currentInfraJob == nil && cluster.Status.ProvisioningJobGeneration != cluster.Generation
+	var expectCreate int
+	if shouldProvision {
+		expectCreate = 1
+	}
+	keysToDelete := []string{}
+	for _, job := range invalidActiveJobs {
+		keysToDelete = append(keysToDelete, jobKey(job))
+	}
+	c.expectations.SetExpectations(key, expectCreate, keysToDelete)
+
+	// Delete invalid jobs asynchronously
+	if len(invalidActiveJobs) > 0 {
+		go c.deleteJobs(key, invalidActiveJobs)
+	}
+
+	// If a current job was found, sync with it
+	if currentInfraJob != nil {
+		cLog.Debugln("provisioning job exists, will sync with job")
+		return c.syncClusterStatusWithJob(cluster, currentInfraJob)
+	}
+
+	// If no current job was found, the cluster is not yet provisioned,
+	// and the status says there should be a job, the
+	// job may have been deleted. Update status accordingly
+	if currentInfraJob == nil && isClusterProvisioning(cluster) && cluster.Status.ProvisioningJobGeneration == cluster.Generation {
+		return c.setJobNotFoundStatus(cluster)
+	}
+
+	if shouldProvision {
+		return c.provisionCluster(cluster)
+	}
+	return nil
+}
+
+func (c *InfraController) deleteJobs(clusterKey string, jobs []*v1batch.Job) {
+	for _, job := range jobs {
+		err := c.kubeClient.BatchV1().Jobs(job.Namespace).Delete(job.Name, &metav1.DeleteOptions{})
+		if err != nil {
+			c.logger.WithField("job", jobKey(job)).Warningf("error deleting job: %v", err)
+			utilruntime.HandleError(err)
+			c.expectations.DeletionObserved(clusterKey, jobKey(job))
```

Comment:
We are not tracking job deletes, so there is no reason to register observations of job deletes.